common/ethereum/applications/lifi/lifi.go

```diff
@@ -0,0 +1,516 @@
+
+	//transactionId := unpacked[0]
+	//dex := unpacked[1]
+	//timestamp := unpacked[6]
+
+	fromAssetId, err := ethereum.CoerceBoundContractResultsToAddress(
+		[]interface{}{unpacked[2]},
+	)
+
+	if err != nil {
+		return feeData, fmt.Errorf(
+			"failed to coerce fromAssetId to address: %v",
+			err,
+		)
+	}
+
+	toAssetId, err := ethereum.CoerceBoundContractResultsToAddress(
+		[]interface{}{unpacked[3]},
+	)
+
+	if err != nil {
+		return feeData, fmt.Errorf(
+			"failed to coerce toAssetId to address: %v",
+			err,
+		)
+	}
+
+	fromAmount, err := ethereum.CoerceBoundContractResultsToRat(
+		[]interface{}{unpacked[4]},
+	)
+
+	if err != nil {
+		return feeData, fmt.Errorf(
+			"failed to coerce fromAmount to *big.Rat: %v",
+			err,
+		)
+	}
+
+	toAmount, err := ethereum.CoerceBoundContractResultsToRat(
+		[]interface{}{unpacked[5]},
+	)
+
+	if err != nil {
+		return feeData, fmt.Errorf(
+			"failed to coerce toAmount to *big.Rat: %v",
+			err,
+		)
+	}
+
+	//timestamp := unpacked[6]
+
+	fluidTransferAmount := new(big.Rat)
+
+	switch fluidTokenContract {
+	case fromAssetId:
+		fluidTransferAmount.Set(fromAmount)
+
+	case toAssetId:
+		fluidTransferAmount.Set(toAmount)
+
+	default:
+		return feeData, fmt.Errorf(
+			"failed to decode the volume: fluid contract address %v is not  asset id (%v) nor is to asset id (%v)",
+			fluidTokenContract,
+			fromAssetId,
+			toAssetId,
+		)
+	}
+
+	decimalsAdjusted := math.Pow10(tokenDecimals)
+	decimalsRat := new(big.Rat).SetFloat64(decimalsAdjusted)
+
+	feeData.Volume = new(big.Rat).Quo(fluidTransferAmount, decimalsRat)
+	feeData.Fee = new(big.Rat).SetInt64(0)
+
+	return feeData, nil
+}
+
+// GetLifiFees unpacks several potential types of topics, and returns the
+// volume. It does not determine the fees paid currently.
+func GetLifiFees(transfer worker.EthereumApplicationTransfer, client *ethclient.Client, fluidTokenContract ethCommon.Address, tokenDecimals int) (feeData applications.ApplicationFeeData, err error) {
+	transferLog := transfer.Log
+
+	if len(transferLog.Topics) < 1 {
+		return feeData, fmt.Errorf("not enough log topics passed!")
+	}
+
+	transferLogData := transfer.Log.Data
+
+	logTopic := transferLog.Topics[0].String()
+
+	// TODO: fees are not being tracked with Lifi
+
+	switch logTopic {
+	case lifiGenericSwapCompletedTopic:
+		unpacked, err := lifiAbi.Unpack("LiFiGenericSwapCompleted", transferLogData)
+
+		if err != nil {
+			return feeData, fmt.Errorf(
+				"failed to unpack a LiFiGenericSwapCompleted log: %v",
+				err,
+			)
+		}
+
+		return genericSwapCompleted(
+			fluidTokenContract,
+			tokenDecimals,
+			unpacked,
+		)
+
+	case lifiSwappedGenericTopic:
+		unpacked, err := lifiAbi.Unpack("LiFiSwappedGeneric", transferLogData)
+
+		if err != nil {
+			return feeData, fmt.Errorf(
+				"failed to unpack a LiFiSwappedGeneric log: %v",
+				err,
+			)
+		}
+
+		return swappedGeneric(
+			fluidTokenContract,
+			tokenDecimals,
+			unpacked,
+		)
+
+	case lifiAssetSwappedTopic:
```

Comment:
will this be emitted from the same addresses as the other topics (therefore crashing the app server?) or is it from a separate contract? or is this meant to use `assetSwapped`?

Reply:
It won't be emitted anyway. If it does, we need to support it with a test. I couldn't find a transaction online to use for a test except on a small Bitcoin shitcoin chain that didn't have an archive node available with the public RPC (the only event was from last year). It seems to be a non supported event that Lifi doesn't emit (though supported via the ABI - with my cursory search).